maps/src/main/java/com/gluonhq/impl/maps/tile/CachedHttpTileRetriever.java

```diff
@@ -0,0 +1,240 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL GLUON BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.gluonhq.impl.maps.tile;
+
+import com.gluonhq.attach.storage.StorageService;
+import com.gluonhq.impl.maps.tile.osm.CachedOsmTileRetriever;
+import com.gluonhq.maps.tile.TileRetriever;
+import javafx.scene.image.Image;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class CachedHttpTileRetriever implements TileRetriever {
+
+    private static final Logger logger = Logger.getLogger(CachedOsmTileRetriever.class.getName());
+
+    private File cacheRoot;
+    private boolean hasFileCache;
+    private CacheThread cacheThread = null;
+
+    private final HttpTileRetriever baseTileRetriever;
+
+    public CachedHttpTileRetriever( HttpTileRetriever baseTileRetriever ) {
+
+        this.baseTileRetriever = Objects.requireNonNull(baseTileRetriever);
+
+        try {
+            File storageRoot = StorageService.create()
+                    .flatMap(StorageService::getPrivateStorage)
+                    .orElseThrow(() -> new IOException("Storage Service is not available"));
+
+            cacheRoot = new File(storageRoot, ".gluonmaps");
+            logger.fine("[JVDBG] cacheroot = " + cacheRoot);
+            if (!cacheRoot.isDirectory()) {
+                hasFileCache = cacheRoot.mkdirs();
+            } else {
+                hasFileCache = true;
+            }
+            if (hasFileCache) {
+                cacheThread = new CacheThread( cacheRoot, baseTileRetriever::buildImageUrlString);
+                cacheThread.start();
+            }
+            logger.info("hasfilecache = " + hasFileCache);
+        } catch (IOException ex) {
+            hasFileCache = false;
+            logger.log(Level.SEVERE, null, ex);
+        }
+    }
+
+    private final static Executor EXECUTOR = Executors.newFixedThreadPool(2, runnable -> {
```

Comment:
2 can be a System.getProperty() that defaults to 2?

Reply:
Yes, that was in the code before I refactored, so I left it unchanged.

Reply:
Updated. It is now retrieved using `gluon.maps.thread.number` system property with the default of `2`